codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java

```diff
@@ -467,6 +436,96 @@ public void settingsWriteShouldNotUpdateSettings() throws Exception {
         verify(writer).writeSettings(eq(ctx), eq(settings), eq(promise));
     }
 
+    @Test
+    public void dataWriteShouldCreateHalfClosedStream() {
+        doAnswer(new Answer() {
+            @Override
+            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
+                FlowControlled flowControlled = (FlowControlled) invocationOnMock.getArguments()[2];
+                flowControlled.write(Integer.MAX_VALUE);
+                flowControlled.writeComplete();
+                return null;
+            }
+        }).when(remoteFlow).sendFlowControlled(eq(ctx), eq(stream), any(FlowControlled.class));
+        ByteBuf data = dummyData();
+        encoder.writeData(ctx, STREAM_ID, data.retain(), 0, true, promise);
+        verify(remoteFlow).sendFlowControlled(eq(ctx), eq(stream), any(FlowControlled.class));
+        verify(lifecycleManager).closeLocalSide(stream, promise);
+        assertEquals(data.toString(UTF_8), writtenData.get(0));
+        data.touch();
+    }
+
+    @Test
+    public void dataWriteShouldHalfCloseStream() throws Exception {
+        doAnswer(new Answer() {
+            @Override
+            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
+                FlowControlled flowControlled = (FlowControlled) invocationOnMock.getArguments()[2];
+                flowControlled.write(Integer.MAX_VALUE);
+                flowControlled.writeComplete();
+                return null;
+            }
+        }).when(remoteFlow).sendFlowControlled(eq(ctx), eq(stream), payloadCaptor.capture());
+        reset(future);
```

Comment:
I didn't write any of this :P I think I accidentally moved it in the code and now it looks like my contribution :D Only added the mocking of the flow controller